pandas/_libs/groupby.pyx

```diff
@@ -686,12 +721,13 @@ def group_mean(floating[:, ::1] out,
     if len_values != len_labels:
         raise ValueError("len(index) != len(labels)")
 
-    nobs = np.zeros((<object>out).shape, dtype=np.int64)
     # the below is equivalent to `np.zeros_like(out)` but faster
+    nobs = np.zeros((<object>out).shape, dtype=np.int64)
     sumx = np.zeros((<object>out).shape, dtype=(<object>out).base.dtype)
     compensation = np.zeros((<object>out).shape, dtype=(<object>out).base.dtype)
 
     N, K = (<object>values).shape
+    nan_val = NPY_NAT if is_datetimelike else NAN
```

Comment:
Somewhat wary of this since `NPY_NAT` is an int so there's some kind of implicit cast here. And large ints (like NPY_NAT) can't be represented losslessly as floats

Reply:
See comment bellow.